### PR TITLE
Fix binary formatter on arrays with non-zero lower bounds

### LINKED ISSUE #16753: Test disabled with "Fails on desktop and core: 'Unable to cast object of type 'System.UInt32[][*]' to type 'System.Object[]'"
Test: Roundtrip_ArrayContainingArrayAtNonZeroLowerBound() under corefx\src\System.Runtime.Serialization.Formatters\tests\BinaryFormatterTests.cs

[ActiveIssue("Fails on desktop and core: 'Unable to cast object of type 'System.UInt32[][*]' to type 'System.Object[]'")]

Detail: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs#L750

### PR DESCRIPTION
**Pull request by alexperovich**
Fixes #16753

This change fixes the problem by casting to Array rather than object[]

### CODE REVIEW

File: src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs [hunk 1]
```diff
@@ -544,10 +544,10 @@ private void WriteArray(WriteObjectInfo objectInfo, NameInfo memberNameInfo, Wri
 
                 _serWriter.WriteJaggedArray(memberNameInfo, arrayNameInfo, arrayElemObjectInfo, arrayElemTypeNameInfo, lengthA[0], lowerBoundA[0]);
 
-                var objectA = (object[])array;
+                var objectA = (Array)array;
                 for (int i = lowerBoundA[0]; i < upperBoundA[0] + 1; i++)
                 {
-                    WriteArrayMember(objectInfo, arrayElemTypeNameInfo, objectA[i]);
+                    WriteArrayMember(objectInfo, arrayElemTypeNameInfo, objectA.GetValue(i));
```
Comment (stephentoub) [inline src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs]: Does this have any measurable impact on the performance of serializing/deserializing arrays?  Or is this only hit for the (very uncommon) case of non-zero-lower-bound arrays?

File: src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs [hunk 2]
```diff
@@ -747,8 +747,8 @@ public void Roundtrip_CrossProcess(object obj)
             }
         }
 
-        [ActiveIssue(16753)] //Fails on desktop and core: 'Unable to cast object of type 'System.UInt32[][*]' to type 'System.Object[]'
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
```
Comment (weshaggard) [inline src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs]: Any reason we cannot make the core behavior match desktop? That is generally the way to go for these as opposed to disabling tests.

Comment (alexperovich) [inline src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs]: Desktop fails in the same way core does today. I can just delete the test if that is preferred, or make it a negative test.

Comment (stephentoub) [inline src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs]: @weshaggard, as noted in the comment on the ActiveIssue, the existing core behavior does match the existing desktop behavior: it fails with an exception "'Unable to cast object of type 'System.UInt32[][*]' to type 'System.Object[]'"

Comment (stephentoub) [inline src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs]: > I can just delete the test if that is preferred, or make it a negative test.

We shouldn't delete the test.  If we're uncomfortable fixing a bug like this, then the answer isn't to delete the test, but rather to not fix the bug.  I'm not sure why we wouldn't fix it, though.

File: src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs (hunk 1 above)
Comment (alexperovich) [inline src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs]: This code path is run when serializing arrays of arrays, and non-zero-lower-bound arrays. When running the entire test project the only things that hit this are `int[][]` and `uint[][*]`

Comment (stephentoub) [inline src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs]: Ok, cool, sounds good.

**Comment by weshaggard** [comment]
@alexperovich I'm still interested in why we don't make the core behavior match that on desktop? As far as I can tell there isn't much if any value in keeping the behavior different.

**Comment by stephentoub** [comment]
@weshaggard, are you asking why we're fixing a bug that's in both core and desktop?

**Comment by weshaggard** [comment]
No I'm asking why we don't make core behave the same as desktop for the test case that we had to add SkipOnFramework(NETFramework). If the behavior is the same I wouldn't expect the SkipOnFramework attribute to be necessary.

**Comment by stephentoub** [comment]
The test is currently marked ActiveIssue and fails without it on both core and desktop. This PR fixes the bug in core and this removes the ActiveIssue from the test as it now passes on core... but it still fails on desktop which still had the bug, so the skip attribute is added for desktop. What am I missing?

**Comment by weshaggard** [comment]
I understand that is what the PR did, but I'm questioning why we fixed it on core as opposed to just keeping the behavior the same as desktop and just negating the test and having no SkipOnFramework for it.

**Comment by stephentoub** [comment]
Why not? What's special about this bug that we wouldn't want to fix it, especially when the fix is so simple? Is your concern specific to BinaryFormatter, that you don't think we should fix bugs in it?

**Comment by weshaggard** [comment]
No my concern is having differences between core and desktop. It is always a red flag to me whenever I see a SkipOnFramework attribute applied to a test because that means we have inconsistent behavior. We should try for consistency unless we feel that there is a strong reason to diverge. We can fix bugs if we feel like we will fix it everywhere eventually, but if that is the case here then we should minimally tag the bug as 'netfx-port-consider`, but I suspect we will never fix this on desktop which is why I would error on the side of consistency.

**Pull request merged**